src/Compiler/pars.fsy

```diff
@@ -3649,6 +3649,47 @@ declExpr:
         let mWhileAll = unionRanges (rhs parseState 1) (rhs parseState 3)
         exprFromParseError (SynExpr.While (spWhile, arbExpr("whileGuard1", mWhileHeader), arbExpr("whileBody3", mWhileBodyArb), mWhileAll))  }
 
+  | WHILE_BANG declExpr doToken typedSequentialExprBlock doneDeclEnd 
+      { let mWhileHeader = unionRanges (rhs parseState 1) $2.Range
+        let spWhile = DebugPointAtWhile.Yes mWhileHeader 
+        let mWhileAll = unionRanges (rhs parseState 1) $4.Range
+        SynExpr.WhileBang (spWhile, $2, $4, mWhileAll) }
+      
+  | WHILE_BANG declExpr doToken typedSequentialExprBlock recover 
+      { if not $5 then reportParseErrorAt (rhs parseState 1) (FSComp.SR.parsUnexpectedEndOfFileWhile())
+        let mWhileHeader = unionRanges (rhs parseState 1) $2.Range
+        let spWhile = DebugPointAtWhile.Yes mWhileHeader 
+        let mWhileAll = unionRanges (rhs parseState 1) $4.Range
+        exprFromParseError (SynExpr.WhileBang (spWhile, $2, $4, mWhileAll)) }
+
+  | WHILE_BANG declExpr doToken error doneDeclEnd 
+      { // silent recovery 
```

Comment:
Just trying to understand - when this (and the other silent recovery) will be triggered?

Reply:
Thanks!